File: src/main/kotlin/com/huawei/cangjie/psi/CjTypeElement.java
package com.huawei.cangjie.psi;

import com.intellij.util.ArrayFactory;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public interface CjTypeElement extends CjElement {


    CjTypeElement[] EMPTY_ARRAY = new CjTypeElement[0];

    ArrayFactory<CjTypeElement> ARRAY_FACTORY = count -> count == 0 ? EMPTY_ARRAY : new CjTypeElement[count];

    // may contain null
    @NotNull
    List<CjTypeReference> getTypeArgumentsAsTypes();
}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjTypeElement.kt
package com.linqingying.cangjie.psi

import com.intellij.util.ArrayFactory

interface CjTypeElement : CjElement {
    val typeArgumentsAsTypes: List<CjTypeReference > get() = emptyList()

    companion object {
        @JvmStatic
        val EMPTY_ARRAY: Array<CjTypeElement?> = arrayOfNulls(0)

        val ARRAY_FACTORY: ArrayFactory<CjTypeElement ?> =
            ArrayFactory { count: Int -> if (count == 0) EMPTY_ARRAY else arrayOfNulls(count) }
    }
}
